### PR TITLE
TP8736 - Make attributes whitelisting optional

### PR DESCRIPTION
**Pull request by baahrens**

### CODE REVIEW

**Comment by gitbot-eve** [comment]
https://compeongmbh.tpondemand.com/entity/8736

File: src/serialize.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ const removeDuplicateIncludes = included => {
 
 const extractResourceInformation = (resource, attributeNames, relationshipNames) => {
   const identifier = extractIdentifier(resource)
-  const permittedAttributes = pick(resource, attributeNames)
+  const permittedAttributes = isEmpty(attributeNames) ? resource : pick(resource, attributeNames)
```
Comment (GerritSe) [inline src/serialize.js:30]: Wdyt about this, so you can explicitly use an empty set of attributes if you want?

```suggestion
  const permittedAttributes = attributeNames == null ? resource : pick(resource, attributeNames)
```

Comment (baahrens) [reply src/serialize.js:30]: That's a nice idea, yea.

File: spec/__snapshots__/serialize.spec.js.snap [hunk 2]
```diff
@@ -56,6 +69,7 @@ Object {
   "data": Object {
     "attributes": Object {
       "firstName": "Nico",
+      "id": "511",
```
[the remark this answers is not in the document]
Comment (baahrens) [reply spec/__snapshots__/serialize.spec.js.snap:72]: Seems to be on purpose tho
Bildschirmfoto 2020-09-03 um 15 46 40

[the remark this answers is not in the document]
Comment (GerritSe) [reply spec/__snapshots__/serialize.spec.js.snap:72]: The id is not supposed to be in `attributes`, but directly in `data`. The snapshots expose that bug.

File: src/serialize.js [hunk 3]
```diff
@@ -25,9 +25,16 @@ const removeDuplicateIncludes = included => {
...
+  const attributes = attributeNames == null ? resource : pick(resource, attributeNames)
+  delete attributes.id
+
+  return attributes
+}
+
 const extractResourceInformation = (resource, attributeNames, relationshipNames) => {
   const identifier = extractIdentifier(resource)
-  const permittedAttributes = pick(resource, attributeNames)
+  const permittedAttributes = pickAttributes(resource, attributeNames)
```
Comment (David3141) [inline src/serialize.js:37]:
```suggestion
  const permittedAttributes = attributeNames == null
    ? omit(resource, ['id'])
    : pick(resource, attributeNames)
```

(or do that in the `pickAttributes` function above)

File: src/serialize.js [hunk 4]
```diff
@@ -27,7 +28,7 @@ const removeDuplicateIncludes = included => {
 
 const extractResourceInformation = (resource, attributeNames, relationshipNames) => {
   const identifier = extractIdentifier(resource)
-  const permittedAttributes = pick(resource, attributeNames)
+  const permittedAttributes = attributeNames == null ? omit(resource, 'id') : pick(resource, attributeNames)
   const polymorphicType = resource[COMPEON_API_JS_TYPE]
   const [relationships, attributes] = partition(permittedAttributes, (_, key) => (
     includes(relationshipNames, key)
```
Comment (David3141) [inline src/serialize.js:34]: Do these lines make sense if we don't want to specify the relationships in the `attributes` anymore?

Comment (baahrens) [reply src/serialize.js:38]: Yes, I think so. We still need to separate attributes and relationships. I did found another error though investigating this. Could you have a look at [this commit](https://github.com/COMPEON/compeon-jsonapi-serializer/pull/194/commits/03a6687fb944de94b970fadfcc7649e8cc5a78d9)?

**Review by David3141 — APPROVED** [review]
Don't forget the fixup 🌚

**Pull request merged**